src/Jackalope/Transport/DoctrineDBAL/Client.php

```diff
@@ -1576,33 +1602,42 @@ private function deleteProperty(string $path): void
         $xpath = new \DOMXPath($dom);
 
         $found = false;
-        $propertyName = PathHelper::getNodeName($path);
-        foreach ($xpath->query(sprintf('//*[@sv:name="%s"]', $propertyName)) as $propertyNode) {
-            $found = true;
-            // would be nice to have the property object to ask for type
-            // but its in state deleted, would mean lots of refactoring
-            if ($propertyNode->hasAttribute('sv:type')) {
+        foreach($paths as $path) {
+            $propertyName = PathHelper::getNodeName($path);
+            $tablesToDeleteReferencesFrom = [];
+            foreach ($xpath->query(sprintf('//*[@sv:name="%s"]', $propertyName)) as $propertyNode) {
+                $found = true;
+                // would be nice to have the property object to ask for type
+                // but its in state deleted, would mean lots of refactoring
+                if (!$propertyNode->hasAttribute('sv:type')) { continue; }
+
                 $type = strtolower($propertyNode->getAttribute('sv:type'));
                 if (in_array($type, ['reference', 'weakreference'])) {
-                    $table = $this->referenceTables['reference' === $type ? PropertyType::REFERENCE : PropertyType::WEAKREFERENCE];
-                    try {
-                        $query = "DELETE FROM $table WHERE source_id = ? AND source_property_name = ?";
-                        $this->getConnection()->executeUpdate($query, [$nodeId, $propertyName]);
-                    } catch (DBALException $e) {
-                        throw new RepositoryException(
-                            'Unexpected exception while cleaning up deleted nodes',
-                            $e->getCode(),
-                            $e
-                        );
-                    }
+                    $tablesToDeleteReferencesFrom[] = $this->referenceTables['reference' === $type ? PropertyType::REFERENCE : PropertyType::WEAKREFERENCE];
```

Comment:
is this an additional optimization to avoid running unncessary deletions in n-m relationships?

Reply:
I wasn't really sure about this one. But the reasoning is the following in the old code it would fire a delete for every instance of the property being found by the xpath selector
```php
$propertyName = PathHelper::getNodeName($path);
$tablesToDeleteReferencesFrom = [];
foreach ($xpath->query(sprintf('//*[@sv:name="%s"]', $propertyName)) as $propertyNode) {
```

However if the query is depending on the `$nodeId` and the `$propertyName` neither of them change in the loop. So running only at most one delete query for a given property seems better.

I don't know if you could simplify it even further and just accumulate references that need to be deleted per node and then delete all of them at the end. (Before flushing the new XML content). It's always a tadeoff between memory and speed.

Reply:
makes sense to me. i think your improvement is good enough.